File: app/src/main/java/com/example/rtspserver/MainActivity.java
package com.example.rtspserver;

import android.Manifest;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.view.SurfaceHolder;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import net.majorkernelpanic.streaming.Session;
import net.majorkernelpanic.streaming.gl.SurfaceView;

import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.util.Enumeration;

public class MainActivity extends AppCompatActivity implements Session.Callback, SurfaceHolder.Callback {

    private final static int PERMISSION_REQUEST_CODE = 1;
    private SurfaceView mSurfaceView;
    private Button mButton;
    private TextView mUrlTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        mSurfaceView = findViewById(R.id.surface);
        mButton = findViewById(R.id.start_stop_button);
        mUrlTextView = findViewById(R.id.rtsp_url);

        // Start the server service once
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(new Intent(this, RtspServer.class));
        } else {
            startService(new Intent(this, RtspServer.class));
        }

        if (!hasPermissions()) {
            requestPermissions();
        } else {
            initializeApp();
        }
    }

    private boolean hasPermissions() {
        return ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED &&
                ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED;
    }

    private void requestPermissions() {
        ActivityCompat.requestPermissions(this,
                new String[]{Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO},
                PERMISSION_REQUEST_CODE);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == PERMISSION_REQUEST_CODE) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED && grantResults[1] == PackageManager.PERMISSION_GRANTED) {
                initializeApp();
            } else {
                Toast.makeText(this, "Permissions not granted. App cannot function.", Toast.LENGTH_LONG).show();
                finish();
            }
        }
    }

    private void initializeApp() {
        mSurfaceView.getHolder().addCallback(this);
        mButton.setOnClickListener(v -> {
            RtspServer server = RtspServerSingleton.getServer();
            if (server != null) {
                Session session = server.getSession();
                if (session != null) {
                    if (session.isStreaming()) {
                        new Thread(() -> session.stop()).start();
                    } else {
                        new Thread(() -> session.start()).start();
                    }
                }
            }
        });
    }

    private String getIpAddress() {
        try {
            for (Enumeration<NetworkInterface> en = NetworkInterface.getNetworkInterfaces(); en.hasMoreElements();) {
                NetworkInterface intf = en.nextElement();
                if (intf.getName().toLowerCase().contains("wlan")) {
                    for (Enumeration<InetAddress> enumIpAddr = intf.getInetAddresses(); enumIpAddr.hasMoreElements();) {
                        InetAddress inetAddress = enumIpAddr.nextElement();
                        if (!inetAddress.isLoopbackAddress() && inetAddress instanceof Inet4Address) {
                            return inetAddress.getHostAddress();
                        }
                    }
                }
            }
        } catch (SocketException e) {
            e.printStackTrace();
        }
        try {
            for (Enumeration<NetworkInterface> en = NetworkInterface.getNetworkInterfaces(); en.hasMoreElements();) {
                NetworkInterface intf = en.nextElement();
                for (Enumeration<InetAddress> enumIpAddr = intf.getInetAddresses(); enumIpAddr.hasMoreElements();) {
                    InetAddress inetAddress = enumIpAddr.nextElement();
                    if (!inetAddress.isLoopbackAddress() && inetAddress instanceof Inet4Address) {
                        return inetAddress.getHostAddress();
                    }
                }
            }
        } catch (SocketException e) {
            e.printStackTrace();
        }
        return "127.0.0.1";
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        stopService(new Intent(this, RtspServer.class));
    }

    // Session.Callback methods
    @Override
    public void onBitrateUpdate(long bitrate) {}

    @Override
    public void onSessionError(int reason, int streamType, Exception e) {
        runOnUiThread(() -> {
            String error = "Session error: " + (e != null ? e.getMessage() : "Unknown error");
            Toast.makeText(MainActivity.this, error, Toast.LENGTH_LONG).show();
            mButton.setText("Start");
        });
    }

    @Override
    public void onPreviewStarted() {}

    @Override
    public void onSessionConfigured() {}

    @Override
    public void onSessionStarted() {
        runOnUiThread(() -> {
            mButton.setText("Stop");
            mUrlTextView.setText("rtsp://" + getIpAddress() + ":8086");
        });
    }

    @Override
    public void onSessionStopped() {
        runOnUiThread(() -> {
            mButton.setText("Start");
            mUrlTextView.setText("RTSP URL: ");
        });
    }

    // SurfaceHolder.Callback methods
    @Override
    public void surfaceCreated(SurfaceHolder holder) {
        RtspServer server = RtspServerSingleton.getServer();
        if (server != null) {
            Session session = server.getSession();
            if (session != null) {
                session.setSurfaceView(mSurfaceView);
                session.setCallback(this);
                session.startPreview();
            }
        }
    }

    @Override
    public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {}

    @Override
    public void surfaceDestroyed(SurfaceHolder holder) {
        RtspServer server = RtspServerSingleton.getServer();
        if (server != null) {
            Session session = server.getSession();
            if (session != null) {
                session.stopPreview();
            }
        }
    }
}

File: app/src/main/java/com/example/rtspserver/MainActivity.kt
package com.example.rtspserver

import android.Manifest
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.SurfaceHolder
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.lifecycle.lifecycleScope
import com.example.rtspserver.databinding.ActivityMainBinding
import com.pedro.common.ConnectChecker
import com.pedro.library.rtsp.RtspCamera2
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity(), ConnectChecker, SurfaceHolder.Callback {

    private lateinit var rtspCamera2: RtspCamera2
    private lateinit var binding: ActivityMainBinding

    private val requestPermissionLauncher =
        registerForActivityResult(
            ActivityResultContracts.RequestMultiplePermissions()
        ) { permissions ->
            if (permissions.all { it.value }) {
                // Permissions granted, start the preview
                rtspCamera2.startPreview()
            } else {
                Toast.makeText(this, "Permissions not granted.", Toast.LENGTH_SHORT).show()
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        rtspCamera2 = RtspCamera2(binding.surfaceView, this)
        // Add the SurfaceHolder.Callback to manage the preview lifecycle
        binding.surfaceView.holder.addCallback(this)

        binding.bStartStop.setOnClickListener {
            lifecycleScope.launch {
                if (!rtspCamera2.isStreaming) {
                    // Te operacje mogą blokować wątek UI, wykonaj je w tle
                    if (rtspCamera2.prepareAudio() && rtspCamera2.prepareVideo()) {
                        binding.bStartStop.text = "Stop Stream"
                        rtspCamera2.startStream("/live/pedro")
                    } else {
                        Toast.makeText(this@MainActivity, "Error preparing stream", Toast.LENGTH_SHORT).show()
                    }
                } else {
                    binding.bStartStop.text = "Start Stream"
                    rtspCamera2.stopStream()
                    binding.tvUrl.text = ""
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()
        // Restart the preview when the activity is resumed and the surface is valid.
        if (hasPermissions() && !rtspCamera2.isOnPreview && binding.surfaceView.holder.surface.isValid) {
            rtspCamera2.startPreview()
        }
    }

    override fun onPause() {
        super.onPause()
        // Stop the stream and preview to release the camera while the app is in the background.
        if (rtspCamera2.isStreaming) {
            rtspCamera2.stopStream()
            binding.bStartStop.text = "Start Stream"
            binding.tvUrl.text = ""
        }
        if (rtspCamera2.isOnPreview) {
            rtspCamera2.stopPreview()
        }
    }

    private fun hasPermissions(): Boolean {
        return arrayOf(Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO).all {
            ActivityCompat.checkSelfPermission(this, it) == PackageManager.PERMISSION_GRANTED
        }
    }

    override fun onConnectionStarted(url: String) {
        runOnUiThread {
            binding.tvUrl.text = url
        }
    }

    override fun onConnectionSuccess() {
        runOnUiThread {
            Toast.makeText(this, "Connection success", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onConnectionFailed(reason: String) {
        runOnUiThread {
            Toast.makeText(this, "Connection failed: $reason", Toast.LENGTH_SHORT).show()
            rtspCamera2.stopStream()
            binding.bStartStop.text = "Start Stream"
            binding.tvUrl.text = ""
        }
    }

    override fun onNewBitrate(bitrate: Long) {
        // Called when bitrate changes
    }

    override fun onDisconnect() {
        runOnUiThread {
            Toast.makeText(this, "Disconnected", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onAuthError() {
        runOnUiThread {
            Toast.makeText(this, "Auth error", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onAuthSuccess() {
        runOnUiThread {
            Toast.makeText(this, "Auth success", Toast.LENGTH_SHORT).show()
        }
    }

    // SurfaceHolder.Callback implementation
    override fun surfaceCreated(holder: SurfaceHolder) {
        // When the surface is created, start the camera preview if permissions are granted
        if (hasPermissions()) {
            rtspCamera2.startPreview()
        } else {
            requestPermissionLauncher.launch(arrayOf(Manifest.permission.CAMERA, Manifest.permission.RECORD_AUDIO))
        }
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        // The library should handle surface changes. Forcing a stop/start here can
        // cause race conditions. If you need to handle orientation changes, you would
        // add more specific logic here, but the previous implementation was unstable.
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        super.surfaceDestroyed(holder)
        // When the surface is destroyed, we must stop the stream and preview.
        // This is a final cleanup step that mirrors onPause.
        if (rtspCamera2.isStreaming) {
            rtspCamera2.stopStream()
        }
        if (rtspCamera2.isOnPreview) {
            rtspCamera2.stopPreview()
        }
    }
}
